embassy-stm32/src/timer/custom_timer.rs

```diff
@@ -0,0 +1,584 @@
+    if_afio!(CustomPwmBuilder<'d, T, CH1, CH2, CH3, ch_mode::Unused, ETR, TS, A>)
+{
+    /// Finalize configuration and create the [CustomPwm]
+    pub fn finalize(self) -> CustomPwm<'d, T> {
+        use ch_mode::Mode;
+        let mut inner = Timer::new(self.tim);
+
+        self.ch1.init(super::Channel::Ch1, &mut inner);
+        self.ch2.init(super::Channel::Ch2, &mut inner);
+        self.ch3.init(super::Channel::Ch3, &mut inner);
+        self.ch4.init(super::Channel::Ch4, &mut inner);
+
+        inner.set_trigger_source(self.trig_source);
+        inner.set_slave_mode(self.slave_mode);
+
+        inner.set_counting_mode(self.counting_mode);
+        match self.speed {
+            Speed::Hertz(hz) => inner.set_frequency(hz),
+            Speed::Manual { arr, psc } => {
+                inner.set_max_compare_value(arr);
+                inner.set_prescaler(psc);
+            }
+        }
+        inner.generate_update_event();
+
+        inner.regs_core().cr1().modify(|r| r.set_opm(self.one_pulse_mode));
+        inner.enable_outputs();
+        inner.start();
+
+        CustomPwm { inner }
+    }
+}
+
+/// Use [CustomPwmBuilder::new] to create a new timer
+pub struct CustomPwm<'d, T: CoreInstance> {
+    inner: Timer<'d, T>,
+}
+
+impl<'d, T: GeneralInstance4Channel> CustomPwm<'d, T> {
+    /// Set compare value
+    pub fn set_compare_value(&mut self, duty: u16, channel: super::Channel) {
+        self.inner.set_compare_value(channel, duty as u32);
+    }
+
+    /// Generate a sequence of PWM waveform
+    ///
+    /// Note:
+    /// you will need to provide corresponding TIMx_UP DMA channel to use this method.
+    pub async fn waveform_up(&mut self, dma: Peri<'_, impl super::UpDma<T>>, channel: super::Channel, duty: &[u16]) {
+        self.inner.waveform_up(dma, channel, duty).await;
+    }
+
+    /// Get capture value for a channel.
+    pub fn get_capture_value(&self, channel: super::Channel) -> u32 {
+        self.inner.get_capture_value(channel)
+    }
+
+    fn new_future(&self, channel: super::Channel) -> InputCaptureFuture<T> {
+        self.inner.enable_channel(channel, true);
+        self.inner.enable_input_interrupt(channel, true);
+
+        InputCaptureFuture {
+            channel,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Asynchronously wait until the pin sees an edge as configured on timer init.
+    // TODO: only allow this to be called for channels in input mode
+    pub async fn wait_for_configured_edge(&mut self, channel: super::Channel) -> u32 {
+        self.new_future(channel).await
+    }
+}
+
+async fn _example(
+    tim: Peri<'_, crate::peripherals::TIM1>,
+    dma: Peri<'_, impl super::UpDma<crate::peripherals::TIM1>>,
+    trigger_pin: crate::peripherals::PA8,
+    out_pin: Peri<'_, crate::peripherals::PA9>,
+    capture_pin: crate::peripherals::PA10,
+) {
+    let out_pin: PwmPin<'_, _, _, crate::gpio::AfioRemap<0>> = PwmPin::new(out_pin, crate::gpio::OutputType::PushPull);
```

Comment:
I can not quite wrap my head around how afio `A` thing is supposed to work. Without specifying the type here, I get an error about it not knowing what type it should be.